pkg/lib/config/workflow.go

```diff
@@ -0,0 +1,597 @@
+	WorkflowAuthenticationMethodSecondaryTOTP        WorkflowAuthenticationMethod = "secondary_totp"
+	WorkflowAuthenticationMethodSecondaryOOBOTPEmail WorkflowAuthenticationMethod = "secondary_oob_otp_email"
+	WorkflowAuthenticationMethodSecondaryOOBOTPSMS   WorkflowAuthenticationMethod = "secondary_oob_otp_sms"
+)
+
+type WorkflowConfig struct {
+	SignupFlows      []*WorkflowSignupFlow      `json:"signup_flows,omitempty"`
+	LoginFlows       []*WorkflowLoginFlow       `json:"login_flows,omitempty"`
+	SignupLoginFlows []*WorkflowSignupLoginFlow `json:"signup_login_flows,omitempty"`
+	ReauthFlows      []*WorkflowReauthFlow      `json:"reauth_flows,omitempty"`
+}
+
+type WorkflowSignupFlow struct {
+	ID    string                    `json:"id,omitempty"`
+	Steps []*WorkflowSignupFlowStep `json:"steps,omitempty"`
+}
+
+type WorkflowSignupFlowStepType string
+
+const (
+	WorkflowSignupFlowStepTypeIdentify     WorkflowSignupFlowStepType = "identify"
+	WorkflowSignupFlowStepTypeAuthenticate WorkflowSignupFlowStepType = "authenticate"
+	WorkflowSignupFlowStepTypeVerify       WorkflowSignupFlowStepType = "verify"
+	WorkflowSignupFlowStepTypeUserProfile  WorkflowSignupFlowStepType = "user_profile"
+)
+
+type WorkflowSignupFlowStep struct {
+	ID   string                     `json:"id,omitempty"`
+	Type WorkflowSignupFlowStepType `json:"type,omitempty"`
+
+	// OneOf is relevant when Type is identify or authenticate.
+	OneOf []*WorkflowSignupFlowOneOf `json:"one_of,omitempty"`
+	// TargetStep is relevant when Type is verify.
+	TargetStep string `json:"target_step,omitempty"`
+	// UserProfile is relevant when Type is user_profile.
+	UserProfile []*WorkflowSignupFlowUserProfile `json:"user_profile,omitempty"`
+}
+
+type WorkflowSignupFlowOneOf struct {
+	// Identification is specific to identify.
+	Identification WorkflowIdentificationMethod `json:"identification,omitempty"`
+
+	// Authentication is specific to authenticate.
+	Authentication WorkflowAuthenticationMethod `json:"authentication,omitempty"`
+	// TargetStep is specific to authenticate.
+	TargetStep string `json:"target_step,omitempty"`
+
+	// Steps are common.
+	Steps []*WorkflowSignupFlowStep `json:"step,omitempty"`
+}
+
+type WorkflowSignupFlowUserProfile struct {
+	Pointer  string `json:"pointer,omitempty"`
+	Required bool   `json:"required,omitempty"`
+}
+
+type WorkflowLoginFlow struct {
+	ID    string                   `json:"id,omitempty"`
+	Steps []*WorkflowLoginFlowStep `json:"steps,omitempty"`
+}
+
+type WorkflowLoginFlowStepType string
+
+const (
+	WorkflowLoginFlowStepTypeIdentify     WorkflowLoginFlowStepType = "identify"
+	WorkflowLoginFlowStepTypeAuthenticate WorkflowLoginFlowStepType = "authenticate"
+)
+
+type WorkflowLoginFlowStep struct {
+	ID   string                    `json:"id,omitempty"`
+	Type WorkflowLoginFlowStepType `json:"type,omitempty"`
+
+	// Optional is relevant when Type is authenticate.
+	Optional *bool `json:"optional,omitempty"`
```

Comment:
Maybe not related to this pr, but what actually is the purpose of having an optional authenticate step?

Reply:
When `optional` is true, and the user does not have any applicable authentication method listed in `one_of`, then this authentication step is skipped. This allows us to represent optional 2FA step. When the user has configured 2FA, then the step must be executed.